### DISCUSSION TITLE
Simultaneous Login via Keycloak - Retrieving the external IDP token response for the current session

### DISCUSSION THREAD

**Discussion by Niki93** [keycloak/keycloak#31596] [Ideas]
**Discussion for an Improvement/ Issue**
 
We are using Keycloak as a broker with the "OpenID Connect v1.0" identity provider and have a use case where we consume [this](https://www.keycloak.org/docs/latest/server_development/#retrieving-external-idp-tokens) Keycloak API to get the external IDP token response.
 
[This API implementation](https://github.com/keycloak/keycloak/blob/6a22f31b812ce68735373b1a0b047b75958f0b64/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L467) returns the latest external IDP token response belonging to the Federated Identity User and not to the specific UserSessionModel
 
Similarly, we are also consuming[ Token Exchange API](https://www.keycloak.org/docs/latest/securing_apps/#token-endpoint) to get a refreshed external IDP access token however it does not return the entire token response that includes claims alongside the access token (see [here](https://github.com/keycloak/keycloak/blob/25.0.2/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L249) and [here](https://github.com/keycloak/keycloak/blob/25.0.2/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L315)).
 
Would anyone had such a use case of simultaneous log in and retrieving the correct token response for the current user session?
 
If you had, what was your workaround/solution/approach?
 
**KC Version:** 23.0.6
 
**Expected Behaviour**
 
Use case:
1. User A is logged in via Browser A and request to get Token Response - Gets Token Response R1
2. User A is logged in via Browser B and request to get Token Response- Gets Token Response R2
3. User A in Browser A re-requests for Token Response -Should get Token Response R1

**Actual Behaviour**
 
In point 3 when User A in Browser A re-requests, Token Response R2 comes back instead.
 
**How to Reproduce**
Use the [this](https://www.keycloak.org/docs/latest/server_development/#retrieving-external-idp-tokens) Keycloak API and try out the expected behaviour

**Comment by danielFesenmeyer** [comment]
As far as I understand, this feature is rather for debugging purposes. It just returns the last IDP token which has been used to login with that IDP. Furthermore, the DB just supports one token per IDP, so this behavior currently seems to be by design.

**Reply by Niki93** [reply]
@danielFesenmeyer Thank you for your reply. :) I have further expanded/clarified my discussion as a comment below.

**Comment by Niki93** [comment]
The [Retrieving external IDP tokens](https://www.keycloak.org/docs/latest/server_development/#retrieving-external-idp-tokens) documentation states:
Application code can retrieve these tokens and responses to pull in extra user information, or to securely invoke requests on the external IDP. For example, an application might want to use the Google token to invoke on other Google services and REST APIs.

The same user logging in for an OIDC Client via 2 different browsers will have 2 UserSessionModels and this can be seen in the KC Admin Console under Sessions

We are trying to get the external IDP token response associated with the current Keycloak session rather than the last retrieved external IDP token response. However that’s currently not the behaviour. 

**Our Usecase**

1. User A is logged in via Browser A for an OIDC client - UserSessionModel#1 is created
2. Gets a Token Response and the tokenReponse#1 is stored in the FederatedIdentityModel 
3. User A is logged in via Browser B for an OIDC client - UserSessionModel#2 is created
4. Gets a Token Response and the tokenReponse#2 is stored/updated in the FederatedIdentityModel
5. User A in Browser A now gets a Token Response again. Gets tokenResponse#2 instead of tokenReponse#1

The token response is also stored as **_FEDERATED_ACCESS_TOKEN_RESPONSE_** in contextData of BrokeredIdentityContext. 


**Proposed Idea/Solution:**

Store the entire external IDP token response as a session note (rather than just the external IDP [access token](https://github.com/keycloak/keycloak/blob/adb2af442a88b88452cd8aee2d064c906d4eb9ec/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L739)) 

- Set FEDERATED_ACCESS_TOKEN_RESPONSE in the AuthenticationSessionModel sessionNotes in method [authenticationFinished](https://github.com/keycloak/keycloak/blob/bf26d3364c58ee884e1a3d551d4ba2d15ecd97db/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L733-L741) so that is available in the UserSessionModel for retrieval.

			
```
try{			
authSession.setUserSessionNote(FEDERATED_ACCESS_TOKEN_RESPONSE, JsonSerialization.writeValueAsString(tokenResponse));
} catch (final IOException e) {
	throw new IdentityBrokerException("Error serializing token response", e);
}
```

@thomasdarimont 
Hoping to hear some reviews/comments on the proposed idea and if the given use case above is valid and is currently being supported?

Thank you!
